### DISCUSSION TITLE
Improvements to Combo Disk Images

### DISCUSSION THREAD

**Discussion by kiwisincebirth** [wwarthen/RomWBW#474] [Ideas]
This was extracted from another discussion, and would be something actioned in future

COMBO (Disk Image) - Add Z3PLUS (full os), either at end as a new slice -or- replace wordstar with Z3PLUS to keep combo image size the same. Main reason ensure the combo imgae has the complete set of OS images

Comments below by @wwarthen 

> There are actually 2 OSes that are not in the Combo image. In addition to Z3PLUS, there is also QDOS. I don't want to change the slice layout of the combo image at this point -- too many people expect the current layout. The layout may need to change someday, but that would be a very big deal and I would need to be convinced that the new layout would not change for a very long time.

> So, I guess the additional OSes could be added to the end of the existing combo image slices. However, this has issues as well:

> Spencer distributes CF Cards that are 64MB. Adding slices to the combo image will cause it to not fit. This is not a real problem because any slices exceeding the end of the media will just be unusable. However, I believe I will get a huge number of support queries when the tools that copy images to media issue ugly warnings/errors. Unfortunately, people don't read the documentation, especially if they have already done this before and expect it to work the same.

> Some people have started using slices after the 6 slices of the combo image for their own data. They have the expectation that appying a new combo image will only overlay the first 6 slices. I have occasionally suggested that people start their own slices at 32 to avoid this in the future, but many have not done this. Applying a larger combo image will destroy their existing data if they are not careful.

> I completely understand that these two issues can easily be worked around by reading documentation and/or use of COPYSL. The problem is not reading documentation.

> I am overdue to generate a new stable release of RomWBW (v3.5). Since quite a few people have already started using the pre-release v3.5, I am inclined to move this change to v3.6. If we make a change to the combo disk layout, it would be best to do it in conjunction with this change (v3.6).

**Comment by kiwisincebirth** [comment]
Hi @wwarthen 

I think should decide what (if any) action to take on this discussion - replacing wordstar with Z3PLUS to keep combo image size the same - is (my) preferred outcome - but clearly may not be for all.

Other option

I had another thought. Why not make it a user preference. I noted your comments in (https://github.com/wwarthen/RomWBW/issues/554) surely this would be simpler since only the file append is affected, don't need to worry about ROM addresses. 

The only issue I see is the Platform config cant be used since this is outside HBIOS, so potentially would need another "config file", used by the script file itself which just enables disables the optional file appends to run.

The other point would be if doing this potentially this builds the existing combo image, OR builds a NEW Combo-2 image, leaving existing combo image as is for backwards compatibility. Combo-2 could be inclusive by default including most images  by default.

But the counter argument is a sophisticated user is going to know how to append files and create there own image and it is well documented process requiring little effort. 

Let me know your thoughts?

**Comment by wwarthen** [comment]
Hi @kiwisincebirth,

I like the concept of allowing a user to define their own custom aggregate image.  Although any user could do this today, it would be easier if there were a simple/standard way to define the contents of a custom image that would be built automatically without writing any command scripts, etc.

- I think the custom image should be in addition to the combo image.  The combo image has become ubiquitous and I am reluctant to repurpose it even if the user does it themselves.

- I don't think it is an issue that this is not defined in the existing configuration file hierarchy.  The images are very intentionally agnostic with respect to the ROM build configurations.  Many users have multiple systems and would most likely want a single custom aggregate disk image for all of them.  So, this would be a separate configuration mechanism.

What I have in mind is allowing the user to create simple files that contain a list of slices.  These "image definition" files would be identified by an extension.  During the build, any files found with the designated extension would be processed to create the associated aggregate image.  By default, there would be none of these files (but perhaps an example).  The user could create multiple such image definition files, if desired.

As an example, the user could create a file called `custom.def`.  When the build runs, it finds the file because it has the extension `.def`.  The file would contain a simple list of the slice images desired.  The build would create `hd1k_custom.img` (and also `hd512_custom.img`) which would be a ready-to-use image including the prefix.  The number of slices in the image would be purely determined by the number of lines in the `.def` file.

A simple `custom.def` file might be:

```
# Leading # denotes a comment line...
# This file defines Wayne's custom disk image
zsdos
hitechc
blank
blank
cpm22
games
cobol
qpm
z3plus
```

The file format could eventually be extended to allow specifying the disk label desired next to each slice specification.

What do you think?

Thanks, Wayne

**Reply by kiwisincebirth** [reply]
@wwarthen I like what you have suggested, here are some thoughts

> By default, there would be none of these files (but perhaps an example). The user could create multiple such image definition files, if desired. As an example, the user could create a file called custom.def. 

To make this more approachable I would suggest that a `custom.def` is included that includes ALL the images, with suitable `#` documentation relating to how the file can be modified, and/or copied and renamed. Then for the simplest of users it just is a case of deleting/commenting lines (images) in the file that don't interest them, and maybe some reordering.

Also you could implement the combo image as a `combo.def` file

**Comment by wwarthen** [comment]
I think we are in sync.  I have opened an enhancement issue for this (Issue #576).  Not sure when I can get to it, but that way I won't lose track of it.

Thanks, Wayne
